api/v1alpha1/dataplane_konnect_extension_types.go

```diff
@@ -0,0 +1,81 @@
+package v1alpha1
+
+/*
+Copyright 2024 Kong Inc.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func init() {
+	SchemeBuilder.Register(&DataPlaneKonnectExtension{}, &DataPlaneKonnectExtensionList{})
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:categories=kong;all
+// +kubebuilder:subresource:status
+
+// DataPlaneKonnectExtension is the Schema for the dataplanekonnectextension API,
+// and is intended to be referenced as extension by the dataplane API.
+// If a DataPlane successfully refers a DataPlaneKonnectExtension, the DataPlane
+// deployment spec gets customized to include the konnect-related configuration.
+type DataPlaneKonnectExtension struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   DataPlaneKonnectExtensionSpec   `json:"spec,omitempty"`
+	Status DataPlaneKonnectExtensionStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// DataPlaneKonnectExtensionList contains a list of DataPlaneKonnectExtension.
+type DataPlaneKonnectExtensionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []DataPlaneKonnectExtension `json:"items"`
+}
+
+// DataPlaneKonnectExtensionSpec defines the desired state of DataPlaneKonnectExtension.
+type DataPlaneKonnectExtensionSpec struct {
+	// ControlPlaneID is the identifier of the Konnect Control Plane.
+	// +kubebuilder:validation:Required
+	ControlPlaneID string `json:"controlPlaneID"`
+
+	// ControlPlaneRegion is the region of the Konnect Control Plane.
+	// If not set, 'us' is used as the default region.
+	// +optional
```

Comment:
How do we convert `ControlPlaneRegion` to API server URL here? Also, the `validation` seems to limit the extension of regions (for example, if `au` is available, customers has to update CRD for using `au`.)
And as @pmalek said in previous comment, this can also be retrieved from `KonnectControlPlane` reference, so a `default` may not be required.

Reply:
The path is built in the following way:
`<CONTROL-PLANE-ID>.<REGION>.tp0|cp0.<URL(tech.konghq.com)>`
- `CONTROL-PLANE-ID` can be either provided or fetched from the CP NamespacedRef, as I just implemented after Patryk's suggestion
- `REGION` has to provided. It could be fetched from the APIAuthConfiguration resource referenced by the CP as it is part of the [ServerURL](https://github.com/Kong/kubernetes-configuration/blob/ac4e44233638b368f3b979c29046519bb2c5f760/config/samples/konnect_controlplane.yaml#L38-L39) section. However, In case the controlplane is referenced via `KonnectID`, we do need to have a way to specify it in the API itself
- `cp0` and `tp0` are hardcoded values, that have to be set depending on the env var to configure
- `URL` has to be specified in this API.

Reply:
I removed the enum on the region, I agree we may want to have it unbounded